### PR TITLE
[FEAT]Switching to merged genesis

### PR DESCRIPTION
**Pull request by parithosh**
- Switches to merge genesis state instead of phase0
- Deprecates mining support
- Updates flags for clients
- Updates keystore type used by lodestar

### CODE REVIEW

File: kurtosis-module/impl/participant_network/cl/lodestar/lodestar_launcher.go
Comment (gbouv) [inline kurtosis-module/impl/participant_network/cl/lodestar/lodestar_launcher.go:214]: I guess all those changes plays a role in the fact that now the clients don't do PoW anymore. But I don't see their version being updated anywhere in this PR, so I'm wondering how that works. Is it because we're pulling `latest` systematically for all of them?

**Review by gbouv — COMMENTED** [review]
It seems that the CI doesn't run on your fork. I'll check that the tests are passing

File: kurtosis-module/impl/participant_network/el/geth/geth_launcher.go
Comment (gbouv) [inline kurtosis-module/impl/participant_network/el/geth/geth_launcher.go:15]: This is independent of your changes i think but this will break on line 237 I believe. If I can push to your branch I can fix that for you

Comment (gbouv) [reply kurtosis-module/impl/participant_network/el/geth/geth_launcher.go]: pushed

**Comment by gbouv** [comment]
Alright so now tests are running, and it seems they are failing. I'll take a look see if I can fix those myself

**Comment by gbouv** [comment]
@parithosh It seems we still have a Beacon client in the Lightouse client. Beacon was the PoS chain before the merge. I assume this can now be removed right? 
It seems it's failing the tests right now at [this line](https://github.com/kurtosis-tech/eth2-merge-kurtosis-module/blob/cd88112987031e75453f0468750e330d4e1b1d9b/kurtosis-module/impl/participant_network/cl/lighthouse/lighthouse_launcher.go#L132) and I'm not sure if it's because the beacon chain is completely gone or if client actually has trouble starting

**Comment by gbouv** [comment]
Interestingly, a very similar issue happens with other CL clients like nimbus. I tried running the module with only this client and here is what I get in its logs:
```
INF 2022-10-04 10:09:21.815+00:00 Launching beacon node                      topics="beacnde" version=v22.9.1-a84545-stateofus bls_backend=BLST cmdParams="@[\"--non-interactive=true\", \"--log-level=DEBUG\", \"--network=/genesis-data/output\", \"--data-dir=/root/consensus-data\", \"--web3-url=http://31.47.144.6:8551\", \"--nat=extip:31.47.144.8\", \"--enr-auto-update=false\", \"--rest\", \"--rest-address=0.0.0.0\", \"--rest-port=4000\", \"--validators-dir=/root/validator-keys\", \"--secrets-dir=/root/validator-secrets\", \"--doppelganger-detection=false\", \"--subscribe-all-subnets=true\", \"--num-threads=4\", \"--jwt-secret=/genesis-data/output/jwtsecret\", \"--metrics\", \"--metrics-address=0.0.0.0\", \"--metrics-port=8008\", \"--subscribe-all-subnets\"]" config="(configFile: None[InputFile], logLevel: \"DEBUG\", logStdout: auto, logFile: None[OutFile], eth2Network: Some(\"/genesis-data/output\"), dataDir: /root/consensus-data, validatorsDirFlag: Some(/root/validator-keys), secretsDirFlag: Some(/root/validator-secrets), walletsDirFlag: None[InputDir], eraDirFlag: None[InputDir], web3Urls: @[\"http://31.47.144.6:8551\"], web3ForcePolling: false, requireEngineAPI: None[bool], nonInteractive: true, netKeyFile: \"random\", netKeyInsecurePassword: false, agentString: \"nimbus\", subscribeAllSubnets: true, slashingDbKind: v2, numThreads: 4, jwtSecret: Some(\"/genesis-data/output/jwtsecret\"), cmd: noCommand, runAsServiceFlag: false, bootstrapNodes: @[], bootstrapNodesFile: , listenAddress: 0.0.0.0, tcpPort: 9000, udpPort: 9000, maxPeers: 160, hardMaxPeers: None[int], nat: (hasExtIp: true, extIp: 31.47.144.8), enrAutoUpdate: false, weakSubjectivityCheckpoint: None[Checkpoint], syncLightClient: false, trustedBlockRoot: None[Eth2Digest], finalizedCheckpointState: None[InputFile], finalizedCheckpointBlock: None[InputFile], nodeName: \"\", graffiti: None[GraffitiBytes], strictVerification: false, stopAtEpoch: 0, stopAtSyncedEpoch: 0, metricsEnabled: true, metricsAddress: 0.0.0.0, metricsPort: 8008, statusBarEnabled: true, statusBarContents: \"peers: $connected_peers;finalized: $finalized_root:$finalized_epoch;head: $head_root:$head_epoch:$head_epoch_slot;time: $epoch:$epoch_slot ($slot);sync: $sync_status|ETH: $attached_validators_balance\", rpcEnabled: None[bool], rpcPort: None[Port], rpcAddress: None[ValidIpAddress], restEnabled: true, restPort: 4000, restAddress: 0.0.0.0, restAllowedOrigin: None[TaintedString], restCacheSize: 3, restCacheTtl: 60, restRequestTimeout: 0, restMaxRequestBodySize: 16384, restMaxRequestHeadersSize: 64, keymanagerEnabled: false, keymanagerPort: 5052, keymanagerAddress: 127.0.0.1, keymanagerAllowedOrigin: None[TaintedString], keymanagerTokenFile: None[InputFile], lightClientDataServe: true, lightClientDataImportMode: only-new, lightClientDataMaxPeriods: None[uint64], inProcessValidators: true, debugForkChoice: false, discv5Enabled: true, dumpEnabled: false, directPeers: @[], doppelgangerDetection: false, syncHorizon: 50, terminalTotalDifficultyOverride: None[TaintedString], validatorMonitorAuto: false, validatorMonitorPubkeys: @[], validatorMonitorTotals: false, safeSlotsToImportOptimistically: None[uint16], suggestedFeeRecipient: None[Address], payloadBuilderEnable: false, payloadBuilderUrl: \"\")"
NOT 2022-10-04 10:09:21.824+00:00 Starting metrics HTTP server               topics="beacnde" url=http://0.0.0.0:8008/metrics
INF 2022-10-04 10:09:21.852+00:00 Threadpool started                         topics="beacnde" numThreads=4
/root/nimbus-eth2/vendor/nim-testutils/testutils/moduletests.nim(21) moduletests
/root/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(2163) main
/root/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(2031) handleStartUpCmd
/root/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(1848) doRunBeaconNode
/root/nimbus-eth2/beacon_chain/nimbus_beacon_node.nim(587) init
/root/nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(22) raiseAssert
/root/nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: Invalid baked-in state: SSZ BeaconState: object dynamic portion starts at invalid offset [AssertionError]
```

So, there's definitely something wrong in the genesis.ssz file. It seems we're using a third part lib to generate it ([this one](https://github.com/protolambda/eth2-testnet-genesis) I believe?)

File: kurtosis-module/impl/module_io/default_params.go
Comment (parithosh) [reply kurtosis-module/impl/module_io/default_params.go:30]: Yea fair enough. But does kurtosis enforce the `pull: always` policy of docker?

File: kurtosis-module/impl/participant_network/cl/lodestar/lodestar_launcher.go
Comment (parithosh) [reply kurtosis-module/impl/participant_network/cl/lodestar/lodestar_launcher.go:214]: The changes for lodestar are related to their release from ~1mo ago, they changed a lot of their flags to follow a similar convention as other clients. So any version that's recent should accept the new flags.

File: kurtosis-module/impl/participant_network/el/geth/geth_launcher.go
Comment (parithosh) [reply kurtosis-module/impl/participant_network/el/geth/geth_launcher.go]: thanks!

**Comment by parithosh** [comment]
So the clients that support the merged genesis are: lighthouse, lodestar, teku. Nimbus and Prysm will error out for now, but will support it in the future. So our testing should be limited to these 3.

**Comment by gbouv** [comment]
Oh okay, then it explains why nimbus is failing. Though lighthouse should be working, and it doesn't seem to be right now. I sent you on discord one genesis.ssz file I downloaded from a failing docker container, can you test it with your local setup and check if it's working?
If it does work, it means something might be wrong with the versions of the clients we're using here. If it doesn't it means something is wrong we how we generate it

**Comment by parithosh** [comment]
Hmm, this is super weird. I can't reproduce it at all :/

My kurtosis version: `0.49.9`

My yaml file:
```
participants:
  - elType: geth
    elImage: ethereum/client-go:v1.10.25
    clType: lighthouse
    clImage: sigp/lighthouse:v3.1.2
network:
  networkId: '3151908'
  depositContractAddress: '0x4242424242424242424242424242424242424242'
  secondsPerSlot: 12
  slotsPerEpoch: 32
  altairForkEpoch: 0
  mergeForkEpoch: 0
  totalTerminalDifficulty: 0
  numValidatorKeysPerNode: 64
  preregisteredValidatorKeysMnemonic: giant issue aisle success illegal bike spike
    question tent bar rely arctic volcano long crawl hungry vocal artwork sniff fantasy
    very lucky have athlete
waitForMining: false
waitForFinalization: true
waitForVerifications: true
verificationsEpochLimit: 5
logLevel: info
```
I've pushed my build as a docker file, there are no diffs to this branch.

```
kurtosis module exec --enclave-id eth2 parithoshj/kurtosis:merged-genesis-x86 --execute-params "$(cat ./merge.yaml)"
```

Give this a shot and let me know, i tested this on an M1 mac (my local dev machine) and a remote machine, it worked fine in both

File: kurtosis-module/impl/module_io/params_deserializer.go
Comment (gbouv) [reply kurtosis-module/impl/module_io/params_deserializer.go:116]: Got it, happy to keep it if it's still needed for certain usecase. Does it mean that we also need to keep the el_client_mining_waiter logic around for besu?

File: kurtosis-module/impl/participant_network/cl/lodestar/lodestar_launcher.go
Comment (gbouv) [reply kurtosis-module/impl/participant_network/cl/lodestar/lodestar_launcher.go:214]: Got it, that makes sense

**Comment by gbouv** [comment]
Okay so I think I understand why the CI (and my previous manual run) were failing. The default params still sets `altairForkEpoch` `mergeForkEpoch ` and `totalTerminalDifficulty ` to non zero value. For now I'll just push a small commit here to set those values to zero by default. Maybe we can chat about what to do here, specifically because besu still needs those right?
On this same note also worth chatting about the waiter you removed. Does besu still needs this as well then? Basically I'm inclined to either:
- we remove the code, and therefore should also remove the config param
- keep the code and the config param, and maybe skip waiting anyway when the 3 above params set to zero
What do you think?

**Review by gbouv — COMMENTED** [review]
Yay all tests pass now. But I guess that's because `waitForVerifications` is set to false in the default params. So still curious to look into why it passes on your laptop VS mine
So, all in all, before merging, I'd like to clarify:
- [ ] Do we still need the waitForMining flag for besu? If yes, then let's keep the miningWaiter I guess?
- [ ] What do we want to do about the `altairForkEpoch`, `mergeForkEpoch` and `totalTerminalDifficulty` params? Keep them for besu? Or remove them? Maybe we should make sure that besu still works here. If it's broken anyway, maybe not worth keeping them and then we can open an issue `Fix besu` to take appropriate actions
- [ ] What do we want to do about the 2 clients that are broken `nimbus` and `prysm`? Should we write somewhere in the docs that they are broken and track in an issue?
- [ ] Try to understand what are those failing tests above, and why they are passing on your machine

**Comment by gbouv** [comment]
Update from our chat:
- We decided to just get rid of all the params related to the merge / PoW. If besu still needs it, they can fork. We also decided to shelf the current master branch to a `legacy/pre-merge` for posterity
- Add a note in the read mentioning nimbus and prysm are broken, with links to the issue
- I'll investigate the errors I'm seeing locally, but it shouldn't block this PR b/c tests are green and it seems to be working on Parithosh laptop

**Review by gbouv — APPROVED** [review]
Approving, if it looks good to you feel free to merge

**Comment by parithosh** [comment]
I just tested it again on my laptop, looks fine on latest head as well. So looks good to merge.

**Pull request merged**
